extension-qr-code/src/org/eclipse/birt/extension/qrcode/QRCodeBuilder.java

```diff
@@ -0,0 +1,179 @@
+	private Label lbDotsWidth;
+	private Label lbEncoding;
+
+	protected QRCodeEditor(Shell shell, QRCodeItem qrItem) {
+		super(shell);
+
+		this.qrItem = qrItem;
+	}
+
+	@Override
+	protected void configureShell(Shell newShell) {
+		super.configureShell(newShell);
+
+		newShell.setText("QRCode Builder"); //$NON-NLS-1$
+	}
+
+	private void openExpression(Text textControl) {
+		String oldValue = textControl.getText();
+
+		ExpressionBuilder eb = new ExpressionBuilder(textControl.getShell(), oldValue);
+		eb.setExpressionProvider(new ExpressionProvider(qrItem.getModelHandle()));
+
+		String result = oldValue;
+
+		if (eb.open() == Window.OK) {
+			result = eb.getResult();
+		}
+
+		if (!oldValue.equals(result)) {
+			textControl.setText(result);
+		}
+	}
+
+	@Override
+	protected Control createDialogArea(Composite parent) {
+		Composite composite = new Composite(parent, SWT.NONE);
+		GridLayout layout = new GridLayout(3, false);
+		layout.marginHeight = convertVerticalDLUsToPixels(IDialogConstants.VERTICAL_MARGIN);
+		layout.marginWidth = convertHorizontalDLUsToPixels(IDialogConstants.HORIZONTAL_MARGIN);
+		layout.verticalSpacing = convertVerticalDLUsToPixels(IDialogConstants.VERTICAL_SPACING);
+		layout.horizontalSpacing = convertHorizontalDLUsToPixels(IDialogConstants.HORIZONTAL_SPACING);
+		composite.setLayout(layout);
+		composite.setLayoutData(new GridData(GridData.FILL_BOTH));
+
+		lbText = new Label(composite, SWT.None);
+		lbText.setText("Text Content:"); //$NON-NLS-1$
+
+		txtText = new Text(composite, SWT.BORDER);
+		GridData gd = new GridData(GridData.FILL_HORIZONTAL);
+		txtText.setLayoutData(gd);
+
+		Button btnExp = new Button(composite, SWT.PUSH);
+		btnExp.setText("..."); //$NON-NLS-1$
+		btnExp.setToolTipText("Invoke Expression Builder");
+
+		btnExp.addSelectionListener(new SelectionAdapter() {
+
+			@Override
+			public void widgetSelected(SelectionEvent event) {
+				openExpression(txtText);
+			}
+		});
+
+		lbDotsWidth = new Label(composite, SWT.None);
+		lbDotsWidth.setText("Width (dots):");
+
+		spDotsWidth = new Spinner(composite, SWT.None);
+		spDotsWidth.setMinimum(21);
+		spDotsWidth.setMaximum(2000);
+		spDotsWidth.setDigits(0);
+		gd = new GridData(GridData.FILL_HORIZONTAL);
+		gd.horizontalSpan = 2;
+		spDotsWidth.setLayoutData(gd);
+
+		lbEncoding = new Label(composite, SWT.None);
+		lbEncoding.setText("Encoding:");
+
+		txtEncoding = new Text(composite, SWT.None);
+		txtEncoding.setLayoutData(gd);
+
+		applyDialogFont(composite);
+
+		initValues();
+
+		return composite;
+	}
+
+	private void initValues() {
+		txtText.setText(qrItem.getText());
+		spDotsWidth.setSelection(qrItem.getDotsWidth());
+		txtEncoding.setText(qrItem.getEncoding());
+	}
+
+	@Override
+	protected void okPressed() {
+
+		try {
+			qrItem.setText(txtText.getText());
+			qrItem.setDotsWidth(Integer.parseInt(spDotsWidth.getText()));
+			qrItem.setEncoding(txtEncoding.getText());
+		} catch (Exception ex) {
+			ex.printStackTrace();
```

Comment:
We should not expect any exceptions here. If some input is invalid - "OK" should be disabled